block-producer/src/batch_builder/mod.rs

```diff
@@ -0,0 +1,205 @@
+use std::{cmp::min, fmt::Debug, sync::Arc, time::Duration};
+
+use async_trait::async_trait;
+use itertools::Itertools;
+use miden_objects::{accounts::AccountId, Digest};
+use tokio::{sync::RwLock, time};
+
+use crate::{block_builder::BlockBuilder, SharedProvenTx, SharedRwVec};
+
+// TRANSACTION BATCH
+// ================================================================================================
+
+/// A batch of transactions that share a common proof. For any given account, at most 1 transaction
+/// in the batch must be addressing that account.
+///
+/// Note: Until recursive proofs are available in the Miden VM, we don't include the common proof.
+pub struct TransactionBatch {
+    txs: Vec<SharedProvenTx>,
+}
+
+impl TransactionBatch {
+    pub fn new(txs: Vec<SharedProvenTx>) -> Self {
+        Self { txs }
+    }
```

Comment:
Not for this PR, but I would do the aggregation in the constructor so that the struct does not actually store transactions but only the summary data (i.e., new accounts, updated accounts, created notes, created nullifiers).

Reply:
Agreed; I put a `+ '_` [constraint on all the types](https://github.com/0xPolygonMiden/miden-node/blob/60691e44717e0407d13c4607ad399a117e399689/block-producer/src/batch_builder/mod.rs#L33) to all for that change in the future. But now that I'm looking at it again, it wouldn't have taken much longer to do it properly right away...